Hashing for constant and singleton SymInt/SymBool

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109171
* __->__ #109170
* #109169

Bugfix:
- previously, SymBool does not implement `__eq__`, Python falls back to default `__eq__ `and `__hash__`
- in this PR, we make SymBool implement `__eq__`
- symbolic SymBool now raises an error when hashed just like SymInt/SymFloat

New feature:
- previously, SymInt and SymFloat are unhashable (even if you are singleton or constant)
- in this PR, SymInt and SymBool are hashable if singleton/constant

Stay the same:
- SymNode are hashable due to default Python behavior

soulitzer committed:
Make singleton SymInt hashable

[ghstack-poisoned]

soulitzer committed:
Update on "Make singleton SymInt hashable"


In this PR:
- Add the ability for SingletonSymInt to be hashed (only from Python). If hash is called on a SymInt that is not singleton, an assertion error is raised. Preserves the preexisting behavior that SymNode can be hashed.

[ghstack-poisoned]

soulitzer committed:
Update on "Make singleton SymInt hashable"


In this PR:
- Add the ability for SingletonSymInt to be hashed (only from Python). If hash is called on a SymInt that is not singleton, an assertion error is raised. Preserves the preexisting behavior that SymNode can be hashed.

[ghstack-poisoned]

soulitzer committed:
Update on "Make singleton SymInt hashable"


In this PR:
- Add the ability for SingletonSymInt to be hashed (only from Python). If hash is called on a SymInt that is not singleton, an assertion error is raised. Preserves the preexisting behavior that SymNode can be hashed.

[ghstack-poisoned]

soulitzer committed:
Update on "Hashing for constant and singleton SymInt/SymBool"


Previously:
- SymNode can already be hashed because __eq__ is not implemented, so Python falls back to some kind of id-based hash. 
- SymBool also has the same id-based hash for the same reason.
- Because SymBool does not implement __eq__ either, equality checks that falls back to the id-based check as well, i.e. `b == True` or `b == c` would be False (even if b, c are both False)
- SymInt raises TypeError "unhashable type"

In this PR:
- Add the ability for SingletonSymInt to be hashed
- Continue to raise TypeError when non-singleton SymInt is hashed
- constant SymBool hashes and performs equality checks as if it were a Python bool (I don't know if we can rely on Python bool always hashing to 0 or 1)
- somewhat (but not completely?) preserve bc for non-constant SymBool in that we still do some kind of id-based hash, but now we use the intrusive ptr's address 
- preserve bc for SymNode hashing, they should still hash the same way as before

[ghstack-poisoned]

soulitzer committed:
Update on "Hashing for constant and singleton SymInt/SymBool"


Previously:
- SymNode can already be hashed because __eq__ is not implemented, so Python falls back to some kind of id-based hash. 
- SymBool also has the same id-based hash for the same reason.
- Because SymBool does not implement __eq__ either, equality checks that falls back to the id-based check as well, i.e. `b == True` or `b == c` would be False (even if b, c are both False)
- SymInt raises TypeError "unhashable type"

In this PR:
- Add the ability for SingletonSymInt to be hashed
- Continue to raise TypeError when non-singleton SymInt is hashed
- constant SymBool hashes and performs equality checks as if it were a Python bool (I don't know if we can rely on Python bool always hashing to 0 or 1)
- preserve existing behavior for non-constant SymBool and SymNode hashing

[ghstack-poisoned]

ezyang commented on torch/__init__.py:
```diff
@@ -384,6 +392,20 @@ def __sym_not__(self) -> "SymBool":
+        if self.node.is_constant():
+            return hash(self.node.bool_())
+        else:
+            return object.__hash__(self)
```
I'd prefer if we throw an error here, as we did previously

soulitzer replied:
Technically we didn't throw an error here previously!
The default behavior for python classes that don't implement __eq__ or __hash__  is to some kind of object identity equality and hashing.
And so, SymBool doesn't error because it does not implement __eq__(why?), but SymFloat and SymInt do error, because they eq IS implemented.
That said, erroring here might be better for symmetry, so that sounds OK to me.

ezyang replied:
That's a bug 🤡

ezyang replied:
Yes if you could fix this up while you are at it that would be great

ezyang commented on torch/__init__.py:
```diff
@@ -384,6 +392,20 @@ def __sym_not__(self) -> "SymBool":
+                return self.node.bool_() == other
+            elif isinstance(other, SymBool) and other.node.is_constant():
+                return self.node.bool_() == other.node.bool_()
+        return self is other
```
No, this is not OK. Equality needs to induce a guard, not do symbolic equality

soulitzer replied:
Going to think about this more, but I think you mean the reverse? AFAIK, we only induce guards when we do `guard_bool`, etc.

soulitzer replied:
Soo yeah, this is preexisting, i.e. `symbool == symbool` has always produced a plain bool since we never implemented `__eq__` on SymBool and the default behavior for eq is to do `is`.

Maybe we could implement `__eq__` for real in SymBool in a follow up?

ezyang replied:
I don't expect fixing it here to have much blast radius. If you do want to do this separately, the fix diff should stack before this one.

ezyang reviewed (changes requested):
symfloat needs to behave symmetrically to symint

soulitzer commented:
> symfloat needs to behave symmetrically to symint

I believe this is already symmetric, the default behavior when (1) __hash__ is NOT implemented, but (2) __eq__ IS implemented, is to give a "type error: unhashable type".

soulitzer committed:
Update on "Hashing for constant and singleton SymInt/SymBool"


Previously:
- SymNode can already be hashed because __eq__ is not implemented, so Python falls back to some kind of id-based hash. 
- SymBool also has the same id-based hash for the same reason.
- Because SymBool does not implement __eq__ either, equality checks that falls back to the id-based check as well, i.e. `b == True` or `b == c` would be False (even if b, c are both False)
- SymInt raises TypeError "unhashable type"

In this PR:
- Add the ability for SingletonSymInt to be hashed
- Continue to raise TypeError when non-singleton SymInt is hashed
- constant SymBool hashes and performs equality checks as if it were a Python bool (I don't know if we can rely on Python bool always hashing to 0 or 1)
- preserve existing behavior for non-constant SymBool and SymNode hashing

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer committed:
Update on "Hashing for constant and singleton SymInt/SymBool"


Bugfix:
- previously, SymBool does not implement `__eq__`, Python falls back to default `__eq__ `and `__hash__`
- in this PR, we make SymBool implement `__eq__`
- symbolic SymBool now raises an error when hashed just like SymInt/SymFloat

New feature:
- previously, SymInt and SymFloat are unhashable (even if you are singleton or constant)
- in this PR, SymInt and SymBool are hashable if singleton/constant

Stay the same:
- SymNode are hashable due to default Python behavior

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"
